### PR TITLE
YAML consolidation prelude: new YAML dump APIs (FR-702)

### PR DESCRIPTION
**Pull request by schopin-pro**
## Description

Introduce a couple of APIs that are needed for the various YAML consolidation efforts. Note that this is technically an API break, but the function that is removed has never been part of a release, so this *should* be OK.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

**Comment by schopin-pro** [comment]
Draft as it depends on #250

File: src/util.c [hunk 1]
```diff
@@ -119,6 +122,157 @@ int find_yaml_glob(const char* rootdir, glob_t* out_glob)
...
+            g_set_error(error, G_MARKUP_ERROR, G_MARKUP_ERROR_INVALID_CONTENT, "Error emitting YAML: %s", emitter->problem);
+            return FALSE;
+            // LCOV_EXCL_STOP
+        }
+    } while (map_count || seq_count);
+    return TRUE;
+}
+
+static gboolean
+drill_into_yaml_subtree(yaml_parser_t *parser, yaml_emitter_t *emitter, char** key, GError** error) {
```
Comment (slyon) [inline src/util.c:162]: This function took me a while to grasp. It all makes sense, but I'd like to see a comment above briefly describing what this is supposed to do. Also stating that `char**keys` is not some kind of string to be returned to the caller (as is usual with this prototype), but rather a vector of the split up prefix.

File: tests/generator/base.py [hunk 2]
```diff
@@ -277,8 +277,6 @@ def validate_generated_yaml(self, yaml_input):
             if os.path.isfile(output_path):
                 with open(output_path, 'r') as generated:
                     output_yaml = yaml.safe_load(generated.read())
-            else:
-                output_yaml = yaml.safe_load('')
```
Comment (slyon) [inline tests/generator/base.py:281]: This change seems kind of unrelated. Could you please clarify why it is part of this commit?

**Review by slyon — COMMENTED** [review]
Thank you Simon! This is another great step in consolidating our YAML parsing story.

It is looking very good overall. I asked for some clarification and mentioned a few nitpicks in the inline comments. We should be good to merge this after those are resolved!

Also, I can confirm that the `netplan_state_write_yaml()` function isn't used anywhere and was never released. So we're good to "break" that API.

File: src/netplan.c [hunk 3]
```diff
@@ -973,16 +964,51 @@ netplan_state_write_yaml(const NetplanState* np_state, const char* file_hint, co
...
+    }
+
+
+    while (iter) {
+        NetplanNetDefinition* netdef = iter->data;
+        to_write = g_list_append(to_write, netdef);
+        iter = iter->next;
+    }
+
+    return netplan_netdef_list_write_yaml(np_state, to_write, out_fd, error);
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c:1011]: Nice catch! I'm fairly sure that if this function leaks its list we'll have many other later on.

I'm wondering if we can't use a decorator to automate g_list_free when we lose the scope, though.

File: tests/generator/base.py (hunk 2 above)
Comment (schopin-pro) [reply tests/generator/base.py:281]: I had to dig into my memory to remember why I changed this one, and I must have been tired when I decided that it would be the good solution: I removed those lines because they weren't hit by code coverage anymore. I didn't bother wondering why, I just removed them.

Now that I've actually thought about it, it means there's a subtle regression: `write_netplan_conf_full()` will now create an empty file if the global state is empty, whereas the previous version wouldn't touch the FS.

There are multiple options there:
1/ we simply don't care
2/ we approximate the previous behavior by removing the created file if it's empty
3/ we keep the exact behavior as before by creating a new function
```c
gboolean
netplan_state_nondefault_globals(NetplanState *np_state)
{
    return (np_state->backend != NETPLAN_BACKEND_NONE)
        || has_openvswitch(&np_state->ovs_settings, NETPLAN_BACKEND_NONE, NULL);
}
```
and using it in `write_netplan_conf_full` to do an early check 

```c
if (!netplan_state_nondefault_globals(&global_state) || !netplan_state_get_netdefs_size(&global_state))
    return;
```

Now that I've done the work, the latter approach is nicer :)

File: src/netplan.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c]: Turns out we don't even need a new list here. Should be fixed in the new patchset.

File: tests/generator/base.py (hunk 2 above)
Comment (schopin-pro) [reply tests/generator/base.py]: Addressed in the latest patchset.

File: src/util.c (hunk 1 above)
Comment (schopin-pro) [reply src/util.c]: I added some comments, renamed the function to emit_yaml_subtree and the argument to yaml_path.

Comment (slyon) [reply src/util.c]: Thank you, I like that much better!

File: tests/generator/base.py (hunk 2 above)
Comment (slyon) [reply tests/generator/base.py]: Empty files can cause issues (e.g. with snapd get/set, as addressed with the most recent netplan.io SRUs in Ubuntu), so thank you for going with option (3) and keeping status quo!

**Review by slyon — APPROVED** [review]
Thank you for addressing the comments, especially the `netplan_state_nondefault_globals()` function should avoid issues with our snapd integration so good idea going with option (3) there!

LGTM. Let's merge this.
